components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func (o *odometry) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	ov := &spatialmath.OrientationVector{Theta: o.orientation.Yaw, OX: 0, OY: 0, OZ: 1}
+	return ov, nil
+}
+
+func (o *odometry) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.orientation.Yaw, nil
+}
+
+func (o *odometry) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.linearVelocity, nil
+}
+
+func (o *odometry) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return geo.NewPoint(o.position.X, o.position.Y), o.position.Z, nil
+}
+
+func (o *odometry) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	return movementsensor.Readings(ctx, o, extra)
+}
+
+func (o *odometry) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (o *odometry) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		LinearVelocitySupported:     true,
+		AngularVelocitySupported:    true,
+		OrientationSupported:        true,
+		PositionSupported:           true,
+		LinearAccelerationSupported: false,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+func (o *odometry) math(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer utils.UncheckedErrorFunc(func() error {
+			o.mu.Lock()
+			defer o.mu.Unlock()
+			return errors.New("math failed")
+		})
+		defer o.activeBackgroundWorkers.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+			}
+			if leftMove || rightMove {
+				// sleep for the desired time interval
+				time.Sleep(time.Duration(o.timeInterval) * time.Millisecond)
+
+				// use GetInParallel to ensure the left and right motors are polled at the same time
+				fs := []rdkutils.FloatFunc{}
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].left.Position(ctx, nil) })
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].right.Position(ctx, nil) })
+
+				_, positions, err := rdkutils.GetInParallel(ctx, fs)
+				if err != nil {
+					o.logger.Error(err)
+				}
+
+				// current position of the left and right motors in revolutions
+				left := positions[0]
+				right := positions[1]
```

Comment:
I thought that because of these lines 
```
for _, f := range fs {
  wg.Add(1)
  go helper(f)
}
```

it would go in order of how the functions were added to `fs`

Reply:
In the third line you quoted, the `go` statement spins up a background goroutine that runs concurrently with everything else. We now lose all guarantees of which one completes first. but here's a way to avoid that:
```go
results = make([]float, len(fs))
func helper(f, i) { // The real one is more complicated; this is just a sketch.
    value, err := f()
    results[i] = value
}

for i, f := range fs {
    wg.Add(1)
    go helper(f, i)
}
```

Now, the nth function writes its result into the nth item in the list.